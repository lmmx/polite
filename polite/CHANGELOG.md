# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1](https://github.com/lmmx/polite/compare/polite-v0.1.0...polite-v0.1.1) - 2025-08-26

### <!-- 1 -->Features

- use a prelude; DF convenience functions ([#7](https://github.com/lmmx/polite/pull/7))

### <!-- 9 -->Other

- quality upgrades ([#8](https://github.com/lmmx/polite/pull/8))

## [0.1.0](https://github.com/lmmx/polite/releases/tag/polite-v0.1.0) - 2025-08-25

### <!-- 4 -->Documentation

- docs rewrites ([#6](https://github.com/lmmx/polite/pull/6))
- *(cli)* update readme with usage ([#3](https://github.com/lmmx/polite/pull/3))

### <!-- 6 -->Testing

- *(cli)* working integration tests ([#2](https://github.com/lmmx/polite/pull/2))

### <!-- 9 -->Other

- connectorx integration ([#5](https://github.com/lmmx/polite/pull/5))
- *(pre-commit)* autofix run
- Update dataframe.rs
- initial commit (cloned from polars-genson repo)
