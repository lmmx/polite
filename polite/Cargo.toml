[package]
authors.workspace = true
categories = ["data-structures", "database"]
description = "rusqlite Ã— Polars bridge library"
documentation = "https://docs.rs/polite/"
edition.workspace = true
homepage.workspace = true
keywords = ["bridge", "dataframe", "polars", "sqlite"]
license.workspace = true
name = "polite"
readme = "README.md"
repository.workspace = true
version = "0.1.2"

[dependencies]
chrono.workspace = true
planus.workspace = true
polars.workspace = true
rusqlite.workspace = true
thiserror.workspace = true

# Vendored
#connectorx.workspace = true
arrow.workspace = true
polars-arrow.workspace = true

anyhow.workspace = true
fehler.workspace = true
itertools.workspace = true
log.workspace = true
owning_ref.workspace = true
rayon.workspace = true
serde_json.workspace = true
sqlparser.workspace = true
url.workspace = true

# Vendored from connectorx: connectorx crate (optional/feature flag deps)
# (On)
fallible-streaming-iterator.workspace = true
r2d2.workspace = true
r2d2_sqlite.workspace = true
rust_decimal.workspace = true
urlencoding.workspace = true

[dev-dependencies]
anyhow.workspace = true
eyre.workspace = true
insta.workspace = true
tempfile.workspace = true

[package.metadata.cargo-machete]
ignored = ["chrono", "planus"]
