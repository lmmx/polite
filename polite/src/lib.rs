//! # Polite: SQLite Ã— Polars Integration
//!
//! A Rust library that provides seamless integration between SQLite databases
//! and Polars DataFrames using ConnectorX for efficient data transfer.
//!
//! ## Features
//!
//! - **Fast data loading**: Use ConnectorX to efficiently load SQLite query results into Polars DataFrames
//! - **Easy data writing**: Save Polars DataFrames directly to SQLite tables
//! - **Type preservation**: Automatic mapping between Polars and SQLite data types
//! - **Simple API**: Minimal boilerplate for common database operations
//!
//! ## Quick Start
//!
//! ```rust
//! use polite::{connect_sqlite, to_dataframe, from_dataframe};
//! use polars::prelude::*;
//!
//! // Open a SQLite connection
//! let conn = connect_sqlite(Some("data.db")).unwrap();
//!
//! // Load data from SQLite into a DataFrame
//! let df = to_dataframe("data.db", "SELECT * FROM users").unwrap();
//!
//! // Save a DataFrame to SQLite
//! from_dataframe(&conn, "new_table", &df).unwrap();
//! ```
//!
//! ## Modules
//!
//! - [`dataframe`] - Functions for converting between DataFrames and SQLite
//! - [`db`] - Database connection utilities

pub mod dataframe;
pub mod db;

// Re-export the main entrypoints at crate root
pub use dataframe::{from_dataframe, to_dataframe};
pub use db::{connect_sqlite, execute_query};

/// Common imports for polite users.
///
/// This prelude module re-exports the most commonly used items from polite,
/// allowing users to quickly import everything they need with a single glob import.
///
/// # Examples
///
/// ```rust
/// use polite::prelude::*;
///
/// // Now you have access to all the main functions:
/// let conn = connect_sqlite(Some("data.db")).unwrap();
/// let df = to_dataframe("data.db", "SELECT * FROM users").unwrap();
/// from_dataframe(&conn, "backup_users", &df).unwrap();
/// ```
pub mod prelude {
    pub use crate::{connect_sqlite, execute_query, from_dataframe, to_dataframe};

    // If you add the convenience functions, include them too:
    pub use crate::{load_dataframe, save_dataframe};
}

/// Create a DataFrame from a SQLite file with error handling and logging.
///
/// This convenience function provides better error messages and optional logging
/// compared to the raw `to_dataframe` function.
///
/// # Arguments
///
/// * `db_path` - Path to the SQLite database file
/// * `sql` - SQL query to execute
///
/// # Examples
///
/// ```rust
/// use polite::load_dataframe;
///
/// let df = load_dataframe("data.db", "SELECT id, name FROM users LIMIT 10")
///     .expect("Failed to load data");
///
/// println!("Loaded {} rows", df.height());
/// ```
pub fn load_dataframe(db_path: &str, sql: &str) -> Result<polars::prelude::DataFrame, String> {
    to_dataframe(db_path, sql)
        .map_err(|e| format!("Failed to load DataFrame from {}: {}", db_path, e))
}

/// Save a DataFrame to SQLite with automatic table creation and better error handling.
///
/// This convenience function handles connection creation and provides clearer error messages.
///
/// # Arguments
///
/// * `db_path` - Path to the SQLite database file (will be created if it doesn't exist)
/// * `table_name` - Name of the table to create/insert into
/// * `df` - The DataFrame to save
///
/// # Examples
///
/// ```rust
/// use polite::save_dataframe;
/// use polars::prelude::*;
///
/// let df = df! {
///     "id" => [1, 2, 3],
///     "name" => ["Alice", "Bob", "Charlie"],
/// }.unwrap();
///
/// save_dataframe("output.db", "users", &df)
///     .expect("Failed to save DataFrame");
/// ```
pub fn save_dataframe(
    db_path: &str,
    table_name: &str,
    df: &polars::prelude::DataFrame,
) -> Result<(), String> {
    let conn = connect_sqlite(Some(db_path))
        .map_err(|e| format!("Failed to connect to {}: {}", db_path, e))?;

    from_dataframe(&conn, table_name, df)
        .map_err(|e| format!("Failed to save DataFrame to table '{}': {}", table_name, e))?;

    Ok(())
}
