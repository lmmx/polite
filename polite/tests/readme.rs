use insta::assert_snapshot;
use polars::prelude::*;
use polite::prelude::*;
use std::fmt::Write;
use tempfile::NamedTempFile;

#[test]
fn test_friends_db() -> anyhow::Result<()> {
    // Open (or create) a SQLite database
    let db = NamedTempFile::new()?;
    let db_path = db.path().to_str().unwrap();
    let conn = connect_sqlite(Some(db_path))?;

    execute_query(&conn, "CREATE TABLE friends_made (id INTEGER, name TEXT)")?;

    let mut out = String::new();

    let nobody = load_dataframe(db_path, "SELECT * FROM friends_made")?;
    writeln!(
        out,
        "ðŸ¤“ I am making friends in SQLite! I don't have any there yet...\n{nobody:?}"
    )?;

    // Create a table to keep your friends' names in
    execute_query(&conn, "INSERT INTO friends_made VALUES (1, 'Alice')")?;
    execute_query(&conn, "INSERT INTO friends_made VALUES (2, 'Bob')")?;
    execute_query(&conn, "INSERT INTO friends_made VALUES (3, 'Charlie')")?;

    // Query your friends back into a Polars DataFrame
    let dbf = to_dataframe(db_path, "SELECT * FROM friends_made")?;
    writeln!(
        out,
        "ðŸª„ I have lovingly restored my friends into a Polars DataFrame:\n{dbf:?}"
    )?;

    // Add some more friends directly from a Polars DataFrame
    let polars_friends = df! {
        "id" => [4_i64, 5],
        "name" => ["Dora", "Eve"],
    }?;

    from_dataframe(&conn, "cool_friends", &polars_friends)?;

    writeln!(
        out,
        "ðŸ†’ My friends from Polars are now my friends in SQLite:\n{polars_friends:?}"
    )?;

    let all_friends = load_dataframe(
        db_path,
        "SELECT * FROM friends_made UNION ALL SELECT * FROM cool_friends ORDER BY id",
    )?;
    writeln!(
        out,
        "ðŸŽ‰ All my friends are politely gathered in a DataFrame:\n{all_friends:?}"
    )?;

    println!("{out}");
    assert_snapshot!(out);

    Ok(())
}
