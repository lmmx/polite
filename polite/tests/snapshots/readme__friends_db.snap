---
source: polite/tests/readme.rs
expression: out
---
🤓 I am making friends in SQLite! I don't have any there yet...
shape: (0, 2)
┌─────┬──────┐
│ id  ┆ name │
│ --- ┆ ---  │
│ i64 ┆ str  │
╞═════╪══════╡
└─────┴──────┘
🪄 I have lovingly restored my friends into a Polars DataFrame:
shape: (3, 2)
┌─────┬─────────┐
│ id  ┆ name    │
│ --- ┆ ---     │
│ i64 ┆ str     │
╞═════╪═════════╡
│ 1   ┆ Alice   │
│ 2   ┆ Bob     │
│ 3   ┆ Charlie │
└─────┴─────────┘
🆒 My friends from Polars are now my friends in SQLite:
shape: (2, 2)
┌─────┬──────┐
│ id  ┆ name │
│ --- ┆ ---  │
│ i64 ┆ str  │
╞═════╪══════╡
│ 4   ┆ Dora │
│ 5   ┆ Eve  │
└─────┴──────┘
🎉 All my friends are politely gathered in a DataFrame:
shape: (5, 2)
┌─────┬─────────┐
│ id  ┆ name    │
│ --- ┆ ---     │
│ i64 ┆ str     │
╞═════╪═════════╡
│ 1   ┆ Alice   │
│ 2   ┆ Bob     │
│ 3   ┆ Charlie │
│ 4   ┆ Dora    │
│ 5   ┆ Eve     │
└─────┴─────────┘
